[FLINK-21127][runtime][checkpoint] Stores finished status for fully finished operators in checkpoint

gaoyunhaii opened the pull request:
## What is the purpose of the change

Stores the information of whether an operator is fully finished into the checkpoint metadata.

To not have to introduce a new version of checkpoint metadata and due to a fully finished operator state always has an empty states, we could stores a special content for the operator state to indicate that the operator is fully finished.

## Brief change log

- 37484933592e8b5a4f5cf83425ac0b7b612c7480 allows set / get fully finished flag for operator state, and marked fully finished operators according to the checkpoint brief computed.


## Verifying this change

This change added unit tests about set and get fully finished flag, and serialization / deserialization of the operator states marked as fully finished.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): **no**
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: **no**
  - The serializers: **no**
  - The runtime per-record code paths (performance sensitive): **no**
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: **no**
  - The S3 file system connector: **no**

## Documentation

  - Does this pull request introduce a new feature? **no**
  - If yes, how is the feature documented? **not applicable**

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/PendingCheckpoint.java:
```diff
@@ -308,6 +316,25 @@ public CompletedCheckpoint finalizeCheckpoint(
+                                            operatorID.getGeneratedOperatorID(),
+                                            jobVertex.getParallelism(),
+                                            jobVertex.getMaxParallelism());
+                            operatorStates.put(operatorID.getGeneratedOperatorID(), operatorState);
+                        }
+
+                        operatorState.markedFullyFinished();
```
IIUC, the actual state should not be present for finished operators for this checkpoint, right?
Can we add a check for this?

Otherwise, we need to discard previous states.

gaoyunhaii replied:
It should be always null, I add the check for it.

rkhachatryan replied:
I think the check is still missing.

gaoyunhaii replied:
Very sorry for miss the check when migrating the code... I re-added the check.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/OperatorState.java:
```diff
@@ -81,10 +93,22 @@ public OperatorID getOperatorID() {
+        checkState(!isFullyFinished());
 
-        if (subtaskIndex < 0 || subtaskIndex >= parallelism) {
+        if (subtaskIndex == FULLY_FINISHED_PLACEHOLDER_INDEX) {
+            markedFullyFinished();
```
Just to clarify, this branch is for deserialization, right?

gaoyunhaii replied:
Yes, this branch is for deserialization. I added comments to this branch

rkhachatryan replied:
Thanks!

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/OperatorState.java:
```diff
@@ -41,6 +41,18 @@
+    private static final int FULLY_FINISHED_PLACEHOLDER_INDEX = -1;
+
+    /** A special content that marks this operator as fully finished. */
+    private static final Map<Integer, OperatorSubtaskState> FULLY_FINISHED_STATE =
+            new HashMap<Integer, OperatorSubtaskState>() {
+                {
+                    this.put(
+                            FULLY_FINISHED_PLACEHOLDER_INDEX,
+                            OperatorSubtaskState.builder().build());
+                }
+            };
```
Why not `Collections.singletonMap(FULLY_FINISHED_PLACEHOLDER_INDEX, OperatorSubtaskState.builder().build()));`?

gaoyunhaii replied:
Very thanks, this is exactly I'm looking for at first, but I do not found it...  :-)

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/PendingCheckpoint.java:
```diff
@@ -308,6 +316,25 @@ public CompletedCheckpoint finalizeCheckpoint(
 
             // make sure we fulfill the promise with an exception if something fails
             try {
+                // Completes the operator state for the fully finished operators
+                for (ExecutionJobVertex jobVertex : checkpointBrief.getFullyFinishedJobVertex()) {
```
nit: extract method?

rkhachatryan reviewed (changes requested):
IIUC, the PR is not finished yet, right?
I couldn't find any usages of `OperatorState.isFullyFinished`.

Besides, I think one check is missing in `PendingCheckpoint`.

gaoyunhaii commented:
Hi Roman @rkhachatryan, very thanks for the review! I addressed the comments and added the missing test in https://github.com/apache/flink/pull/14754/commits/0f8924698701b1d6768f8757414d69836b83fa3a. 

The usage of the  `fullyFinished` flag is indeed not include in this PR. The flag is expected to be passed to the task after recovery in [FLINK-21088](https://issues.apache.org/jira/browse/FLINK-21088), and used to skip the execution of the fully finished operator in [FLINK-21089](https://issues.apache.org/jira/browse/FLINK-21089). In detail, it would make the fully finished source directly send EndOfPartitionEvent to finish the channel, and skip lifecycle methods like `open`/`close` for the fully finished operators.

rkhachatryan commented:
Hi @gaoyunhaii , thanks for clarifying and updating the PR.
It LGTM in general, I'd make a final pass after all conflicts resolved.

aimiebell commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java:
```diff
@@ -1095,7 +1084,7 @@ public void initFailureCause(Throwable t) {
     void vertexFinished() {
         assertRunningInJobMasterMainThread();
-        final int numFinished = ++verticesFinished;
+        final int numFinished = ++numFinishedVertices;
```
I detect that this code is problematic. According to the [Correctness (CORRECTNESS)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#correctness-correctness), [DLS: Overwritten increment (DLS_OVERWRITTEN_INCREMENT)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#dls-overwritten-increment-dls-overwritten-increment).
The code performs an increment operation (e.g., i++) and then immediately overwrites it. For example, i = i++ immediately overwrites the incremented value with the original value.

gaoyunhaii commented:
Hi Roman @rkhachatryan , I rebased this PR to the latest master, and also have some changes in the `get*` methods of the `OperatorState` to avoid the finished flag affects the normal processes. Could you have another look~? Very thanks!

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/OperatorState.java:
```diff
@@ -118,14 +142,30 @@ public ByteStreamStateHandle getCoordinatorState() {
         return Collections.unmodifiableMap(operatorSubtaskStates);
     }
 
     public Collection<OperatorSubtaskState> getStates() {
+        if (isFullyFinished()) {
+            return Collections.emptySet();
+        }
```
I'm not sure that this check is needed. If `isFullyFinished` then `subtaskStates` should be empty anyways. So continuing wouldn't hurt but the check makes me feel that it's **necessary** in **all** methods.

ditto: other methods

Probably a subclass `FinishedOperatorState` with always empty state would make it easier and less error-prone, WDYT?

gaoyunhaii replied:
I think it would indeed make it more clear and not worry about consistency between methods. I updated the PR with this method. Very thanks!

rkhachatryan reviewed (changes requested):
@gaoyunhaii
Hi Yun Gao, thanks for rebasing the PR.

I don't see you previous changes addressing my comments. Could you check whether they were pushed?

gaoyunhaii commented:
Very thanks @rkhachatryan for the reviewing and very sorry for missing the commits again... I update the PR according to the comments and also include the previous ones.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/FullyFinishedOperatorState.java:
```diff
@@ -0,0 +1,43 @@
+ * A special operator state implementation representing the operators whose instances are all
+ * finished.
+ */
+public class FullyFinishedOperatorState extends OperatorState {
```
Should this class also override `equals`/`hashCode`?

And why do we allow to set coordinator state (i.e. don't override like `putState`)?
Can finished vertex have coordinator state?

gaoyunhaii replied:
We should indeed need to override equals / hashcode so we could distinguish the finished operator and the empty operator state.  I added these methods.

Regarding coordinator state, currently I also do not have specific examples that keep coordinator state after all the operators are finished. For fully finished operators, we would directly skip their execution after failover, thus all its state should not take effect. Now I also disabled setting the coordinator state. Very thanks for pointing out these issues !

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/metadata/MetadataV3Serializer.java:
```diff
@@ -103,11 +105,16 @@ protected void serializeOperatorState(OperatorState operatorState, DataOutputStr
-        for (Map.Entry<Integer, OperatorSubtaskState> entry : subtaskStateMap.entrySet()) {
-            dos.writeInt(entry.getKey());
-            serializeSubtaskState(entry.getValue(), dos);
+        if (operatorState.isFullyFinished()) {
+            dos.writeInt(-1);
```
nit: to me, explicitly writing `isFullyFinished` would be more clear and less error-prone:
```
dos.writeBoolean(operatorState.isFullyFinished());
dos.writeInt(subtaskStateMap.size()); // zero
// empty loop
```

gaoyunhaii replied:
I'm a bit hesitate here since we should not be able to add new field to the meta data since it would impact the compatibility. There would be the cases that user taking a savepoint or external checkpoint with an older version (like 1.12) and resume the savepoint/checkpoint with new version (like the 1.13) to do the upgrade. If we add new fields, we should not e able to parse the existing savepoint / checkpoint ?

dawidwys replied:
I see @gaoyunhaii concerns. I'd be fine with using `-1` here.

rkhachatryan reviewed (changes requested):
Thanks for updating your PR @gaoyunhaii 
I've noticed that build failures **are** related to PR changes (test code).
I've also left a couple of comments, PTAL.

gaoyunhaii commented:
Hi Roman @rkhachatryan , very thanks for the review! I updated the  PR to address the comments. And very sorry for not pay attention to the result... Now I fixed the issue via adding the `serialVersionUID`.

dawidwys commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/FullyFinishedOperatorState.java:
```diff
@@ -0,0 +1,80 @@
+ */
+public class FullyFinishedOperatorState extends OperatorState {
+
+    private static final long serialVersionUID = -7094972830573632176L;
```
Please use `1L` for `serialVersionUID` as recommended in our coding guidelines: https://flink.apache.org/contributing/code-style-and-quality-java.html

dawidwys commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/metadata/MetadataV2Serializer.java:
```diff
@@ -71,6 +73,10 @@ public CheckpointMetadata deserialize(
             throws IOException {
+        checkState(
+                !operatorState.isFullyFinished(),
+                "Could not support finished Operator state in stale serializers.");
```
```suggestion
                "Could not support finished Operator state in state serializers.");
```

dawidwys commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/PendingCheckpointTest.java:
```diff
@@ -544,6 +546,69 @@ public void testDisposeDisposesCoordinatorStates() throws Exception {
+        TaskAcknowledgeResult result =
+                pendingCheckpoint.acknowledgeTask(
+                        runningTaskId, taskStateSnapshot, new CheckpointMetrics(), null);
+        assertEquals(TaskAcknowledgeResult.SUCCESS, result);
```
nit: We prefer `assertThat` as it gives a nicer description in case of a failure.

dawidwys reviewed (approved):
Had just some minor nits.

gaoyunhaii commented:
Hi @dawidwys  very thanks for the review! I have updated the PR according to the comments~

dawidwys commented:
Thanks for the update. Merging.
